### PR TITLE
feat: admin endpoints sso migration

### PR DESCRIPTION
**Pull request by pawelpaszki**
## Description
KFM changes (without connector related changes) for admin API endpoints authorization migration to internal RH SSO

## Verification Steps
1. Add a cluster (can be a non-existent one) to the `config/dataplane-cluster-configuration.yaml`, so that kafka can be created in the database
2. Change the `config/admin-authz-configuration.yaml` so that a rover group that you are member of is present in the configuration (this will enable you accessing admin API endpoints)
3. Create a kafka (it will be in failed state without using existing OSD cluster, it doesn't matter)
4. login to the db container (`make db/login`) and change `organisation_id` to one that you are not a member of, e.g. 

`update kafka_requests set organisation_id = '12147054';`

5. use rhoas-cli to get access token:

- Download https://github.com/redhat-developer/app-services-cli
- `mkdir ~/rhoas-config`
- `export RHOASCONFIG=~/rhoas-config/config.json`
- `export AUTHURL=https://auth.redhat.com/auth/realms/EmployeeIDP`
- `rhoas login --auth-url=$AUTHURL`
- `rhoas authtoken`

6. Confirm that access to the admin API endpoints is granted by hitting `http://localhost:8000/api/kafkas_mgmt/v1/admin/kafkas` with the token obtained in step 5.

## Checklist (Definition of Done)

- [x] All acceptance criteria specified in JIRA have been completed
- [ ] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment) - partially (connectors tests are broken)
- [x] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] Required metrics/dashboards/alerts have been added (or PR created).
- [ ] Required Standard Operating Procedure (SOP) is added.
- [ ] JIRA has been created for changes required on the client side

### CODE REVIEW

File: config/admin-authz-configuration.yaml
Comment (akoserwal) [inline config/admin-authz-configuration.yaml:11]: @R-Lawton Do you agree with role name and formate?

Comment (R-Lawton) [reply config/admin-authz-configuration.yaml:11]: Yeah the kas fleet manager ones look good, for the connectors why don't we differentiate the environment?

Comment (pawelpaszki) [reply config/admin-authz-configuration.yaml:11]: there is more to do on the connectors side (including the roles names), but I guess I can change it as well

**Comment by pawelpaszki** [comment]
@dhirajsb @lburgazzoli @valdar - this PR is required to migrate admin API endpoints to use internal RH SSO. it breaks connectors tests. I would appreciate if you could review it and create a separate PR for the connectors required changes or help me with making those changes

**Review by pb82 — APPROVED** [review]
Verified locally, getting a request from the Admin API using the rhoas auth token.

File: pkg/client/keycloak/config.go
Comment (machi1990) [inline pkg/client/keycloak/config.go:107]: Could this be named something like
```
AdminAPISSORealm
```

and make it configurable? Otherwise this really makes it hard to configure for someone looking to use something different

File: pkg/client/keycloak/config.go
Comment (machi1990) [inline pkg/client/keycloak/config.go:108]: Probably this won't be needed if https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1293/files#r979956339 is addressed

File: config/admin-authz-configuration.yaml
Comment (machi1990) [inline config/admin-authz-configuration.yaml:25]: Maybe we could stick with the previous values as default values

Ref: the discussion we've via DM

File: templates/service-template.yml
Comment (machi1990) [reply templates/service-template.yml:653]: Once the change is done on our side, the lines:
- https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/fd1700778497d649ae4943d8d732c664f6361e28/internal/kafka/internal/environments/production.go#L18 and
- https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/fd1700778497d649ae4943d8d732c664f6361e28/internal/kafka/internal/environments/stage.go

File: internal/kafka/internal/environments/integration.go
Comment (machi1990) [inline internal/kafka/internal/environments/integration.go:48]: lets add the below flags
```
            - --admin-api-sso-base-url=http://127.0.0.1:8180
            - --admin-api-sso-endpoint-uri=/auth/realms/rhoas-kafka-sre
            - --admin-api-sso-realm=rhoas-kafka-sre
```

File: internal/kafka/internal/environments/development.go
Comment (machi1990) [reply internal/kafka/internal/environments/development.go:39]: same change has to be done for the sister file in connectors code base

File: internal/kafka/internal/environments/stage.go
Comment (machi1990) [inline internal/kafka/internal/environments/stage.go:17]:
```suggestion
```
to be set via service template

File: internal/kafka/internal/environments/integration.go
Comment (machi1990) [inline internal/kafka/internal/environments/integration.go:48]:
```suggestion
		
```
not needed as it is already set here https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/fd1700778497d649ae4943d8d732c664f6361e28/pkg/auth/roles_authz.go#L38

File: internal/kafka/internal/environments/development.go
Comment (machi1990) [inline internal/kafka/internal/environments/development.go:39]:
```suggestion
```
not needed as it is set via https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/fd1700778497d649ae4943d8d732c664f6361e28/pkg/auth/roles_authz.go#L38

File: pkg/auth/roles_authz.go
Comment (machi1990) [inline pkg/auth/roles_authz.go:59]:
```suggestion
func (c *AdminRoleAuthZConfig) ReadFiles() error {
	if c.Enabled {
		if err := readRoleAuthZConfigFile(c.RolesConfigFile, &c.RolesConfig); err != nil {
			return err
		}
	}
	return nil
}

```

This method does two things:
- it reads the file 
- it also perform the validation

It is possible to remove the validation method into a separate method and make the config as a `ServiceValidator()`

in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/fd1700778497d649ae4943d8d732c664f6361e28/pkg/providers/core.go#L42
it will read 
```go

di.Provide(auth.NewAdminAuthZConfig, di.As(new(environments.ConfigModule)), di.As(new(environments.ServiceValidator))),
```

And to perform the validation, define a `Validate` method for the `AdminRoleAuthZConfig`  config with the below signature
```go
func (c *AdminRoleAuthZConfig) Validate(env *environments.Env) error {
	if c.Enabled {
		if err := readRoleAuthZConfigFile(c.RolesConfigFile, &c.RolesConfig); err != nil {
			return err
		}
		return validateRolesConfiguration(c.RolesConfig)
	}
	return nil
}
```

File: pkg/providers/core.go
Comment (machi1990) [inline pkg/providers/core.go:42]: let's also register this as ServiceValidator as per https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1293/files#r980230464 comment

File: docs/admin-api-endpoints-authorization.md
Comment (machi1990) [inline docs/admin-api-endpoints-authorization.md:1]: Love this. Thanks for adding this.

File: docs/admin-api-endpoints-authorization.md
Comment (machi1990) [inline docs/admin-api-endpoints-authorization.md:4]: Let's also include these https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/fd1700778497d649ae4943d8d732c664f6361e28/templates/service-template.yml#L419-L429 in the documentation

File: Makefile
Comment (machi1990) [inline Makefile:805]: Thanks for updating the make file.
Can we also update the parameter section under this documentatio https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/fd1700778497d649ae4943d8d732c664f6361e28/docs/deploying-kas-fleet-manager-to-openshift.md#deploy-kas-fleet-manager

File: pkg/auth/roles_authz.go
Comment (machi1990) [inline pkg/auth/roles_authz.go:88]:
```suggestion
```

File: internal/kafka/internal/handlers/authentication.go
Comment (machi1990) [inline internal/kafka/internal/handlers/authentication.go:34]:
```suggestion
			KeysURL(KeycloakConfig.AdminAPISSORealm.JwksEndpointURI). 
```

File: templates/service-template.yml
Comment (machi1990) [inline templates/service-template.yml:178]:
```suggestion
  value: "[{method: GET, roles: ['kas-fleet-manager-admin-full', 'kas-fleet-manager-admin-read', 'kas-fleet-manager-admin-write']}, {method:PATCH, roles: ['kas-fleet-manager-admin-full', 'kas-fleet-manager-admin-write']}, {method:DELETE, roles: ['kas-fleet-manager-admin-full']}]"
```

We could get rid of connectors configuration as this has to be configured on their template as well @valdar @dhirajsb

File: templates/service-template.yml
Comment (machi1990) [inline templates/service-template.yml:955]:
```suggestion
              mountPath: /config/admin-authz-configuration.yaml
```

File: templates/service-template.yml
Comment (machi1990) [inline templates/service-template.yml:1144]:
```suggestion
            - --admin-api-sso-realm=${ADMIN_API_SSO_REALM}
            - --admin-authz-config-file=/config/admin-authz-configuration.yaml
```

File: internal/kafka/internal/routes/route_loader.go
Comment (machi1990) [inline internal/kafka/internal/routes/route_loader.go:258]: Why was this middleware removed?

**Review by machi1990 — CHANGES_REQUESTED** [review]
Thanks @pawelpaszki 

I've left a couple other feedback.

File: templates/service-template.yml
Comment (machi1990) [reply templates/service-template.yml:178]: @valdar @dhirajsb @lburgazzoli can you coordinate this on the connectors side?

File: pkg/auth/roles_authz.go
Comment (machi1990) [inline pkg/auth/roles_authz.go:29]: This value is not used anywhere outside this file, is the intention to enable / disable roles authz check based on this variable? If so, should this be used in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/4b2405a45f1444cf3b591c1cf75c6f04872e6d44/pkg/auth/roles_authz_middleware.go#L42-L57 as well, and only perform the roles check when this variable is set to `true`. 

If not, I think we can remove this boolean flag altogether.

File: templates/service-template.yml
Comment (dhirajsb) [reply templates/service-template.yml:178]: @machi1990 I'm taking a look at all the changes we need to get this to work for connectors.

Comment (machi1990) [reply templates/service-template.yml:178]: Thanks @dhirajsb 

> 'm taking a look at all the changes we need to get this to work for connectors.

@pawelpaszki will be able to help with that

File: internal/kafka/internal/routes/route_loader.go
Comment (pawelpaszki) [reply internal/kafka/internal/routes/route_loader.go:258]: do you think we should temporarily keep it here?

File: pkg/auth/roles_authz.go
Comment (pawelpaszki) [reply pkg/auth/roles_authz.go:29]: will remove the flag. thanks

File: internal/kafka/internal/routes/route_loader.go
Comment (machi1990) [reply internal/kafka/internal/routes/route_loader.go:258]: Yes, I think the `AuditLogMiddleware` configuration should be kept there. 
```
	adminRouter.Use(auth.NewAuditLogMiddleware().AuditLog(errors.ErrorNotFound))
```

Comment (pawelpaszki) [reply internal/kafka/internal/routes/route_loader.go:258]: oh. sorry. good catch!

File: templates/service-template.yml
Comment (machi1990) [reply templates/service-template.yml:955]: All the configs are located in the root `/config`.
This suggestion adds a missing `/`

File: pkg/auth/roles_authz.go
Comment (machi1990) [inline pkg/auth/roles_authz.go:51]:
```suggestion
	return readRoleAuthZConfigFile(c.RolesConfigFile, &c.RolesConfig)
```
one small suggestion

**Review by machi1990 — APPROVED** [review]
the change lgtm, thanks a lot @pawelpaszki for the great work.

let's wait for the connectors team to be get the gist of what config change they'll need to perform on their OpenShift deployment template and AppInterface, before we can merge. /cc @dhirajsb

**Comment by dhirajsb** [comment]
I've pushed fixes for connectors to the PR branch. The connectors integration tests pass locally and can be verified with the command:
```
make test/integration/connector
```

**Comment by dhirajsb** [comment]
I've figured the config changes we require for connectors in app-interface. But if you can share links to kas-fleet-manager PRs for config changes it would help with a sanity check for us.

**Comment by machi1990** [comment]
> I've pushed fixes for connectors to the PR branch. The connectors integration tests pass locally and can be verified with the command:

Thanks @dhirajsb 

> I've figured the config changes we require for connectors in app-interface. But if you can share links to kas-fleet-manager PRs for config changes it would help with a sanity check for us.

I'll let @pawelpaszki share the links to AppInterface privately but one such changes (and the required one) are the ones in service-template.yaml file: https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1293/files#diff-6ae4b2c9b570b88a66bbad92366432c7928538e1238cde45cc4e48cb6fe40df8, I guess those will be needed in https://github.com/bf2fc6cc711aee1a0c2a/cos-fleet-manager/blob/main/templates/service-template.yml

File: templates/service-template.yml
Comment (dhirajsb) [inline templates/service-template.yml:652]: Shouldn't this be `admin-authz-configuration.yaml`?

**Review by dhirajsb — CHANGES_REQUESTED** [review]

**Comment by dhirajsb** [comment]
> I guess those will be needed in https://github.com/bf2fc6cc711aee1a0c2a/cos-fleet-manager/blob/main/templates/service-template.yml

The cos-fleet-manager service changes are in the draft PR https://github.com/bf2fc6cc711aee1a0c2a/cos-fleet-manager/pull/171

File: templates/service-template.yml
Comment (pawelpaszki) [reply templates/service-template.yml]: thanks. fixed

**Review by dhirajsb — APPROVED** [review]
lgtm

File: Makefile
Comment (machi1990) [inline Makefile:801]:
```suggestion
deploy/service: ADMIN_AUTHZ_CONFIG ?= "[{method: GET, roles: [kas-fleet-manager-admin-full, kas-fleet-manager-admin-read, kas-fleet-manager-admin-write]}, {method: PATCH, roles: [kas-fleet-manager-admin-full, kas-fleet-manager-admin-write]}, {method: DELETE, roles: [kas-fleet-manager-admin-full, kas-fleet-manager-admin-write]}]"
```

File: docs/deploying-kas-fleet-manager-to-openshift.md
Comment (machi1990) [inline docs/deploying-kas-fleet-manager-to-openshift.md:195]:
```suggestion
- `ADMIN_AUTHZ_CONFIG`: Configuration file containing endpoints and roles mappings used to grant access to admin API endpoints, Defaults to`"[{method: GET, roles: [kas-fleet-manager-admin-full, kas-fleet-manager-admin-read, kas-fleet-manager-admin-write]}, {method: PATCH, roles: [kas-fleet-manager-admin-full, kas-fleet-manager-admin-write]}, {method: DELETE, roles: [kas-fleet-manager-admin-full, kas-fleet-manager-admin-write]}]
"`
```

File: templates/service-template.yml
Comment (machi1990) [inline templates/service-template.yml:178]:
```suggestion
  value: "[{method: GET, roles: [kas-fleet-manager-admin-full, kas-fleet-manager-admin-read, kas-fleet-manager-admin-write]}, {method: PATCH, roles: [kas-fleet-manager-admin-full, kas-fleet-manager-admin-write]}, {method: DELETE, roles: [kas-fleet-manager-admin-full, kas-fleet-manager-admin-write]}]"
```

**Review by machi1990 — CHANGES_REQUESTED** [review]
Suggestion to add correct default value in Makefile for OpenShift deployment

File: templates/service-template.yml
Comment (machi1990) [reply templates/service-template.yml:178]: @pawelpaszki this needs to be applied as well

**Review by machi1990 — COMMENTED** [review]
lgtm

let's remove the `[DO NOT MERGE]` whenever we are ready to merge before merging :-D

**Review by machi1990 — APPROVED** [review]

**Pull request merged**
